openag/cli/firmware/base.py

```diff
@@ -237,16 +248,28 @@ def all_pio_dependencies(self):
         return deps
 
     def all_git_dependencies(self):
-        deps = set()
+        '''
+        Return a list of all git dependencies.
+        Each dependency is a dict containing a url, type and (optional) branch
+        and tag field:
+
+            [{'url', 'type', 'branch'?, 'tag'?}]
+        '''
+        deps = []
         for plugin in self.plugins:
-            deps = deps.union(plugin.git_dependencies())
+            deps = deps + plugin.git_dependencies()
         for mod_info in self.modules.values():
-            if mod_info.get("repository", {}).get("type", None) == "git":
-                deps.add(mod_info["repository"]["url"])
+            repo_dep = mod_info.get("repository", {})
+            if repo_dep.get("type", None) == "git":
+                deps.append(repo_dep)
             for mod_dep in mod_info.get("dependencies", []):
-                if mod_dep["type"] == "git":
-                    deps.add(mod_dep["url"])
-        return deps
+                if mod_dep.get("type", None) == "git":
+                    deps.append(mod_dep)
+        # De-dupe by indexing each dependency by url. Then read de-duped values
+        # as list.
+        # @TODO in future, we should support same repo at multiple revisions.
```

Comment:
Supporting different versions of the same repo is going to be pretty hard. For now, we could just check if there were any duplicates (`len(unique_deps) != len(deps)`) and raise a warning if so